plutus-conformance/add-test-output/Spec.hs

```diff
@@ -0,0 +1,111 @@
+
+{-# LANGUAGE OverloadedStrings #-}
+
+module Main
+    ( main
+    ) where
+
+import Common (UplcProg, evalUplcProg, stripePosProg)
+import Control.Exception (SomeException, evaluate, try)
+import Data.Text qualified as T
+import Options.Applicative
+import PlutusCore.Error (ParserErrorBundle (ParseErrorB))
+import PlutusCore.Evaluation.Result (EvaluationResult (..))
+import PlutusCore.Pretty (Pretty (pretty), Render (render))
+import PlutusCore.Quote (runQuoteT)
+import System.Directory (doesFileExist)
+import Test.Tasty.Golden (findByExtension)
+import Text.Megaparsec (errorBundlePretty)
+import UntypedPlutusCore.Parser as UPLC (parse, program)
+
+-- |  The arguments to the executable. (1) file extension to be searched,
+-- (2) directory to be searched,
+-- (3) the action to run the input files through; eval (for evaluation tests) or typecheck (for typechecking tests).
+data Args = MkArgs String FilePath Runner
+
+ext :: Parser String
+ext =
+  strArgument
+  (metavar "EXT" <>
+    help "The input file(s) with this extension will be included." )
+
+dir :: Parser FilePath
+dir =
+  strArgument
+    (metavar "DIR" <> help "The directory the input files are in." )
+
+data Runner =
+  Eval
+  | Typecheck
+  deriving stock (Show)
+
+runner :: Parser Runner
+runner = argument
+  (eitherReader runnerReader)
+  (metavar "RUNNER" <> help "The action to apply to the input files that generate the outputs. Either eval or typecheck." )
+
+runnerReader :: String -> Either String Runner
+runnerReader "eval" = Right Eval
+runnerReader "typecheck" = Right Typecheck
+runnerReader inp =
+  Left ("Unsupported test " <> show inp <>
+        ". Please choose either eval (for evaluation tests) or typecheck (for typechecking tests).")
+
+args :: ParserInfo Args
+args = info ((MkArgs <$> ext <*> dir <*> runner) <**> helper)
+  (fullDesc <> progDesc helpText)
+
+helpText :: String
+helpText = "This program adds test outputs to specified inputs." <>
+  "To run the program, input the following 3 arguments: " <>
+  "(1) file extension to be searched " <>
+  "(2) directory to be searched " <>
+  "(3) the action to run the input files through; eval (for evaluation tests) or typecheck (for typechecking tests). " <>
+  "E.g. run " <>
+  "`cabal run add-test-output .uplc plutus-conformance/uplc/ eval` " <>
+  "to have the executable search for files with extension `.uplc` in the /uplc directory that are missing output files. " <>
+  " It will evaluate and create output files for them."
+
+
+main :: IO ()
+main = do
+    MkArgs extension directory run <- customExecParser (prefs showHelpOnEmpty) args
+    allInputFiles <- findByExtension [extension] directory
+    -- only choose the ones without an output file, so as to not edit the ones already with outputs
+    inputFiles <- sequenceA $
+      [do
+              hasOut <- doesFileExist (testIn <> ".expected")
+              if hasOut then pure [] else pure [testIn] | testIn <- allInputFiles]
```

Comment:
I had to stare at it for like 10 seconds to figure out how it even parses. Can it be just

```haskell
inputFiles <- filterM (\testIn -> doesFileExist testIn <> ".expected") allInputFiles
```

?

BTW, instead of

```haskell
if hasOut then pure [] else pure [testIn]
```

you can write

```haskell
pure [testIn | hasOut]
```

Reply:
Also, for functions like `filterM` the witherable package is very useful: https://hackage.haskell.org/package/witherable